### PR TITLE
Fix function clash for non-storage arguments

### PR DESCRIPTION
**Pull request by Amxx**
This is caused by `getFunctionId` not resolving the UDVT. 

Given the following solidity code:

```solidity
library UdvtConflict {
    type myFirstType is bytes32;
    type mySecondType is bytes32;

    function unwrap(myFirstType t) internal pure returns (bytes32) {
        return myFirstType.unwrap(t);
    }

    function unwrap(mySecondType t) internal pure returns (bytes32) {
        return mySecondType.unwrap(t);
    }
}
```

`getFunctionId` will return 
- `unwrapUdvtConflict.myFirstType`
- `unwrapUdvtConflict.mySecondType`

The system doesn't see them as colliding, when in fact they do collide.

### CODE REVIEW

**Comment by frangio** [comment]
The issue was more general as only storage arguments were previously handled, which are converted to `uint256` indices and thus are particularly likely to cause conflicts. Other types like contract types can also cause conflicts and were not handled correctly.

I've generalized the fix so that we get the "ABI type" of every argument, which in the case of UDVT is the underlying.

File: src/core.ts [hunk 1]
```diff
@@ -419,7 +422,7 @@ function isTypeExternalizable(typeName: TypeName | null | undefined, deref: ASTD
   if (typeName == undefined) {
     return true;
   } if (typeName.nodeType === 'UserDefinedTypeName') {
```
Comment (Amxx) [inline src/core.ts:424]:
```suggestion
  } else if (typeName.nodeType === 'UserDefinedTypeName') {
```

Comment (Amxx) [reply src/core.ts:424]: not necessary, but cleaner

File: src/core.ts [hunk 2]
```diff
@@ -282,13 +282,16 @@ function areFunctionsFullyImplemented(contract: ContractDefinition, deref: ASTDe
...
 }
 
-function getFunctionNameQualified(fn: FunctionDefinition, context: ContractDefinition, deref: ASTDereferencer, onlyStorage: boolean = true): string {
-  return fn.name + (onlyStorage ? getStorageArguments(fn, context, deref) : getFunctionArguments(fn, context, deref))
+function getFunctionNameQualified(fn: FunctionDefinition, context: ContractDefinition, deref: ASTDereferencer, onlyConflicting: boolean): string {
+  let args = getFunctionArguments(fn, context, deref);
+  if (onlyConflicting) {
+    args = args.filter(a => a.type !== a.abiType || a.storageType !== undefined);
+  }
+  return fn.name + args
```
Comment (Amxx) [inline src/core.ts:294]:
```suggestion
  return fn.name + getFunctionArguments(fn, context, deref)
    .filter(a => !onlyConflicting || a.type !== a.abiType || a.storageType !== undefined)
```

File: src/core.ts [hunk 3]
```diff
@@ -508,6 +515,50 @@ function getType(typeName: TypeName, context: ContractDefinition, deref: ASTDere
...
+      const { typeString } = typeName.typeDescriptions;
+      assert(typeString != undefined);
+      return typeString;
+
+    case 'UserDefinedTypeName':
+      const typeDef = derefUserDefinedTypeName(deref, typeName);
+      switch (typeDef.nodeType) {
+        case 'UserDefinedValueTypeDefinition':
+          const { underlyingType } = typeDef;
+          const { typeString } = underlyingType.typeDescriptions;
```
Comment (Amxx) [inline src/core.ts:538]:
```suggestion
          const { typeString } = typeDef.underlyingType.typeDescriptions;
```

File: src/core.ts (hunk 2 above)
Comment (frangio) [reply src/core.ts:294]: I find the current code easier to reason about.

**Pull request merged**